### DISCUSSION TITLE
Unable to find widget through query_one inside a function on a different screen

### DISCUSSION THREAD

**Discussion by maxigaz** [Textualize/textual#2612] [Q&A] [answered]
I'm using Textual 0.25.0. I have two screens with each having a single `Static` widget for displaying text. If I press Enter on `ScreenOne`, I would like to run a function declared on `ScreenTwo` that updates the widget on `ScreenTwo` (namely `TextTwo`).

The problem is I can't identify `TextTwo` through `query_one`, because when I do so, the program ends with a `NoMatches` error. I've also noticed that according to `app.log(app.tree)` called inside the function, `TextTwo` is not even listed as part of the DOM tree, even though `ScreenTwo` is pushed already.

I'm not sure if I'm even trying to tackle this with the right approach, so any advice would be very welcome.

Here's my example code:

```python
from textual.app import App, ComposeResult
from textual import events
from textual.screen import Screen
from textual.widgets import Static

class TextOne(Static):
    pass

class TextTwo(Static):
    pass

class ScreenOne(Screen):
    def compose(self) -> ComposeResult:
        yield TextOne("Press Enter to update text on ScreenTwo through a function.")

    def on_key(self, event: events.Key) -> None:
        if event.key == "enter":
            app.push_screen("two")
            app.query_one("ScreenTwo").displayinfo()

class ScreenTwo(Screen):
    def displayinfo(self):
        """A function that does something complicated and displays the result."""
        app.log(app.tree)
        app.query_one("TextTwo").update("New text on ScreenTwo.")

class App(App):
    SCREENS = {"one": ScreenOne(), "two": ScreenTwo()}

    def on_mount(self) -> None:
        app.push_screen("one")

app = App()
app.run()
```

**Comment by TomJGooding** [comment]
Your `TextTwo` class is never used in your example app above?

**Reply by maxigaz** [reply]
Right, I forgot to create it with `yield`.

```python
def compose(self) -> ComposeResult:
    yield TextTwo("")
```

If I add this right under `class ScreenTwo(Screen):`, I still get the same error though.

**Reply by TomJGooding** [reply]
Why not simply call your `display_info` method in `ScreenTwo.on_mount`? I think there might be some other context missing here, perhaps you could explain in a bit more detail what it is you are trying to do?

**Reply by maxigaz** [reply]
> Why not simply call your `display_info` method in `ScreenTwo.on_mount`?

You are right, that also works. Thank you!

> I think there might be some other context missing here, perhaps you could explain in a bit more detail what it is you are trying to do?

I've been working on an application that allows one to display status information about a Syncthing instance.

It currently has two main parts:

1. Two input fields where the user can enter a URL and an API key to the Syncthing instance.
1. A Markdown widget where info should appear.

My goal is to separate these two parts by placing each of them on a screen. So, after the user has filled out the input fields, pressing Enter would process the content of the fields (by sending requests using Syncthing's REST API) and then display the Markdown widget in place of the input fields.

[Here](https://codeberg.org/maxigaz/syncthing-term-status/src/commit/4d14e177428d3f82915c0f8fbef9a3e4ed025bf1/main.py) is the original code without using screens at all and [here](https://codeberg.org/maxigaz/syncthing-term-status/src/commit/f47f7284eec4f5bec7b87e84b31c5c9bffc6f5c5/main.py) is my attempt at rewriting it with the use of screens.

As with the example I posted in my question, I also keep running  into the `NoMatches` error. What seems to be problematic according to the traceback is having a `watch_word` inside the second screen, which runs the `self.query_one("StatusDisplay").update(self.word)` whenever the reactive variable `word` changes. I'm yet to figure out how to resolve this.

Does this help with understanding what I'm trying to do?

**Reply by davep** [reply]
Without diving deep into your code (so forgive me if I've misread it), you're liking getting the `NoMatches` error because queries are restricted to a single screen. More generally, if you're wanting to do something in one screen, get the result and then use it in another screen, you're probably going to want to use the [screen callback facility](https://textual.textualize.io/guide/screens/#returning-data-from-screens).

**Reply by maxigaz** [reply, marked as answer]
I've managed to figure it out eventually.

As I mentioned in my previous reply, I can confirm that calling `display_info` from `on_mount` works in the minimal example.

Now, in my Syncthing project, I had to cut `query_one` from the `watch` method and (still within the second screen) move it inside `on_mount`. I still don't understand fully why this is needed, but I no longer receive the error.

Anyway, thank you for your help!

**Reply by davep** [reply]
Normally if you're getting that error from a watch method it's because you're doing something to trigger the watch before the DOM is ready.
